diff -ur a/src/Plugin/Field/FieldWidget/InlineParagraphsWidget.php paragraphs/src/Plugin/Field/FieldWidget/InlineParagraphsWidget.php
--- a/src/Plugin/Field/FieldWidget/InlineParagraphsWidget.php	2020-05-21 04:35:06.000000000 -0700
+++ paragraphs/src/Plugin/Field/FieldWidget/InlineParagraphsWidget.php	2020-06-11 11:01:16.500563158 -0700
@@ -652,10 +652,10 @@
         if (method_exists(FormatterHelper::class, 'formProcess')) {
           $element['subform']['#process'][] = [FormatterHelper::class, 'formProcess'];
         }
-        elseif (function_exists('field_group_form_pre_render')) {
+        if (function_exists('field_group_form_pre_render')) {
           $element['subform']['#pre_render'][] = 'field_group_form_pre_render';
         }
-        elseif (function_exists('field_group_form_process')) {
+        if (function_exists('field_group_form_process')) {
           $element['subform']['#process'][] = 'field_group_form_process';
         }
       }
Only in paragraphs/src/Plugin/Field/FieldWidget: InlineParagraphsWidget.php.orig
Only in paragraphs/src/Plugin/Field/FieldWidget: InlineParagraphsWidget.php.rej
diff -ur a/src/Plugin/Field/FieldWidget/ParagraphsWidget.php paragraphs/src/Plugin/Field/FieldWidget/ParagraphsWidget.php
--- a/src/Plugin/Field/FieldWidget/ParagraphsWidget.php	2020-05-21 04:35:06.000000000 -0700
+++ paragraphs/src/Plugin/Field/FieldWidget/ParagraphsWidget.php	2020-06-11 10:50:52.893040068 -0700
@@ -770,10 +770,10 @@
         if (method_exists(FormatterHelper::class, 'formProcess')) {
           $element['subform']['#process'][] = [FormatterHelper::class, 'formProcess'];
         }
-        elseif (function_exists('field_group_form_pre_render')) {
+        if (function_exists('field_group_form_pre_render')) {
           $element['subform']['#pre_render'][] = 'field_group_form_pre_render';
         }
-        elseif (function_exists('field_group_form_process')) {
+        if (function_exists('field_group_form_process')) {
           $element['subform']['#process'][] = 'field_group_form_process';
         }
       }
Only in paragraphs/src/Plugin/Field/FieldWidget: ParagraphsWidget.php.orig
Only in paragraphs/src/Plugin/Field/FieldWidget: ParagraphsWidget.php.rej
